z3/src/z3_solver.cpp

```diff
@@ -85,12 +86,74 @@ const unordered_map<PrimOp, variadic_fun> z3_variadic_ops({
 
 void Z3Solver::set_opt(const std::string option, const std::string value)
 {
-  throw NotImplementedException("Set opt not implemented for Z3 backend.");
+  const char * o = option.c_str();
+  const char * v = value.c_str();
+
+  // READ PLEASE
+  // The easiest handling of Z3's set function's param requirements is to have
+  // vectors with the names of different options in the list correspoinding with
+  // which param the z3 api expects, it's worth discussing what options we think
+  // should go in these lists to start and obviously it is very easy to add more
+  // down the line
+  vector<string> bool_opts = { "produce-models", "produce-proofs" };
+  vector<string> string_opts = {};
+  vector<string> int_opts = {};
+
+  if (option == "incremental")
+  {
+    if (value == "false")
+    {
+      throw IncorrectUsageException(
+          "Z3 backend is always incremental -- it cannot be disabled.");
+    }
+  }
+  else if (std::find(bool_opts.begin(), bool_opts.end(), option)
+           != bool_opts.end())
+  {
+    if (value == "true")
+    {
+      slv.set(o, true);
+    }
+    else if (value == "false")
+    {
+      slv.set(o, false);
+    }
+    else
+    {
+      throw IncorrectUsageException("Expected a boolean value.");
+    }
+  }
+  else if (std::find(string_opts.begin(), string_opts.end(), option)
+           != string_opts.end())
+  {
+    slv.set(o, v);
+  }
+  else if (std::find(int_opts.begin(), int_opts.end(), option)
+           != int_opts.end())
+  {
+    try
+    {
+      double num = stoi(value, nullptr, 10);
+      slv.set(o, num);
+    }
+    catch (z3::exception & err)
+    {
+      throw IncorrectUsageException("Expected an integer value.");
+    }
+  }
+  else
+  {
+    std::string msg("Option - ");
+    msg += option;
+    msg += " - not implemented for Z3 backend.";
+    throw NotImplementedException(msg.c_str());
+  }
 }
 
 void Z3Solver::set_logic(const std::string logic)
 {
-  throw NotImplementedException("Set logic not implemented for Z3 backend.");
+  const char * l = logic.c_str();
+  slv = solver(ctx, l);
```

Comment:
So this is Z3-specific question. Do you need to do anything to clean up the old solver?

I think this is good, but in case it's relevant I implemented `set_logic` in mathsat which has a similar API with https://github.com/makaimann/smt-switch/pull/203. I could have reconstructed it like you're doing here, but it seems like mathsat leaks memory if you reconstruct, so instead I waited to construct the env (equivalent of the solver) until after all options / logic is set. A little heavy-handed, but technically works.